skt/__init__.py

```diff
@@ -503,24 +508,53 @@ def mktgz(self, clean=True):
 
         logging.info("building kernel: %s", args)
 
-        mk = subprocess.Popen(args,
-                              stdout=subprocess.PIPE,
-                              stderr=subprocess.STDOUT)
-        (stdout, stderr) = mk.communicate()
-        for line in stdout.split("\n"):
-            m = re.match("^Tarball successfully created in (.*)$", line)
-            if m:
-                tgzpath = m.group(1)
-                break
-
-        fpath = None
-        if tgzpath is not None:
-            fpath = "/".join([self.path, tgzpath])
+        with io.open(self.buildlog, 'wb') as writer, \
+                io.open(self.buildlog, 'rb') as reader:
+            make = subprocess.Popen(args,
+                                    stdout=writer,
+                                    stderr=subprocess.STDOUT)
+
+            timer = Timer(timeout, make.terminate)
+            # If the process is cancelled, it doesn't need to wait for anything
+            timer.setDaemon(True)
+            timer.start()
+            while make.poll() is None:
+                self.append_and_log2stdout(reader.readlines(), stdout_list)
+                time.sleep(1)
+            self.append_and_log2stdout(reader.readlines(), stdout_list)
+            if not timer.isAlive():
+                raise CommandTimeoutError(
+                    "'%s' was taking too long" % ' '.join(args)
+                )
```

Comment:
It seems to me this can race with the timer handler (`make.terminate`) and produce an error even if `make` has finished on time. For example, if it finished during that `time.sleep(1)` or during blocked output to stdout. Could you perhaps write a timer handle which runs `make.poll()` and sets a variable if `make` was still alive on timeout, and *then* calls `make.terminate()`? Then you can check that variable and throw an error if it's set.

BTW, shouldn't we also be checking make's exit status?

Reply:
Yes, it wasn't checking the returncode. I have added some test cases that should validate the correct behavior. Please take a look at it again.

Reply:
Thanks a lot :smiley:! However, timer could still have expired right after make failed, but before we checked, and then we would report the wrong error. Do you think you can fix that?

Reply:
oh, I didn't get it before. Good catch! Please take a look at it now.